tests/Doctrine/Tests/DBAL/Functional/StatementTest.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+namespace Doctrine\Tests\DBAL\Functional;
+
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Types\Type;
+
+class StatementTest extends \Doctrine\Tests\DbalFunctionalTestCase
+{
+    public function testStatementIsReusableAfterClosingCursor()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_reusable');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_reusable', array('id' => 1));
+        $this->_conn->insert('stmt_test_reusable', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_reusable ORDER BY id');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+
+        $stmt->closeCursor();
+
+        $stmt->execute();
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(1, $id);
+        $id = $stmt->fetchColumn();
+        $this->assertEquals(2, $id);
+    }
+
+    public function testClosedCursorDoesNotContainResults()
+    {
+        $sm = $this->_conn->getSchemaManager();
+        $table = new Table('stmt_test_no_results');
+        $table->addColumn('id', 'integer');
+        $sm->createTable($table);
+        $this->_conn->insert('stmt_test_no_results', array('id' => 1));
+        $this->_conn->insert('stmt_test_no_results', array('id' => 2));
+
+        $stmt = $this->_conn->prepare('SELECT id FROM stmt_test_no_results');
+
+        $stmt->execute();
+
+        $id = $stmt->fetchColumn();
+        $this->assertNotEmpty($id);
+
+        $stmt->closeCursor();
+
+        try {
+            $value = $stmt->fetchColumn();
+        } catch (\Exception $e) {
+            // some adapters trigger PHP error or throw adapter-specific exception in case of fetching
+            // from a closed cursor, which still proves that it has been closed
```

Comment:
@morozov I don't have a strong opinion about either of both options. I am just concerned about what the actual behaviour should be when fetching without having executed the query before. Because IMO the whole API defined in `Doctrine\DBAL\Driver\ResultStatement` is in theory only available after executing the prepared statement. I think of it as if one would return a `Result` object (which is in fact a cursor) form the `execute()` method. So to say an encapsulation of the result cursor. I think APIs like JDBC have similar constucts. So in this case the `$result` flag would make more sense to me as it is explicitly set when having a result rather than having as `$isCursorClosed` which does not have a meaning as long as we the prepared statment is not executed. I think I saw something like the `$result` flag in some driver already...